validator/lib/jsonapi/validator/resource.rb

```diff
@@ -0,0 +1,123 @@
+require 'jsonapi/validator/document'
+require 'jsonapi/validator/relationship'
+
+module JSONAPI
+  module Validator
+    class Resource
+      # Validate the structure of a resource create/update payload. Optionally
+      #   validate whitelisted/required id/attributes/relationships, as well as
+      #   primary type and relationship types.
+      #
+      # @param [Hash] document The input JSONAPI document.
+      # @param [Hash] params Validation parameters.
+      #   @option [Hash] permitted Permitted attributes/relationships/primary
+      #     id. Optional. If not supplied, all legitimate fields are permitted.
+      #   @option [Hash] required Required attributes/relationships/primary id.
+      #     Optional. The fields must be explicitly permitted, or permitted
+      #     should not be provided.
+      #   @option [Hash] types Permitted primary/relationships types. Optional.
+      #     The relationships must be explicitly permitted. Not all
+      #     relationships' types have to be specified.
+      # @raise [JSONAPI::Validator::InvalidDocument] if document is invalid.
+      #
+      # @example
+      #   params = {
+      #     permitted: {
+      #       id: true, # Simply ommit it to forbid it.
+      #       attributes: [:title, :date],
+      #       relationships: [:comments, :author]
+      #     },
+      #     required: {
+      #       id: true,
+      #       attributes: [:title],
+      #       relationships: [:comments, :author]
+      #     },
+      #     types: {
+      #       primary: [:posts],
+      #       relationships: {
+      #         comments: {
+      #           kind: :has_many,
+      #           types: [:comments]
+      #         },
+      #         author: {
+      #           kind: :has_one,
+      #           types: [:users, :superusers]
+      #         }
+      #       }
+      #     }
+      #   }
+      def self.validate!(document, params = {})
+        Document.ensure!(document.is_a?(Hash),
+                         'A JSON object MUST be at the root of every JSONAPI ' \
+                         'request and response containing data.')
+        Document.ensure!(document.key?('data') || !document['data'].is_a?(Hash),
```

Comment:
I believe `data` can be an `Array` when POSTing to a relationship endpoint http://jsonapi.org/format/#crud-updating-to-many-relationships